File: src/main/java/com/campool/service/UserService.java
package com.campool.service;

import com.campool.encrypt.Encryptor;
import com.campool.exception.NoSuchUserException;
import com.campool.mapper.UserMapper;
import com.campool.model.UserInfo;
import com.campool.model.UserSignUpRequest;
import com.campool.model.UserUpdateRequest;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor
@Service
public class UserService {

    @NonNull
    private final UserMapper userMapper;

    @NonNull
    private final Encryptor encryptor;

    public void add(UserSignUpRequest userSignUpRequest) {
        if (isDuplicate(userSignUpRequest.getId())) {
            throw new DuplicateKeyException("중복된 아이디가 존재합니다.");
        } else {
            userMapper.insertUser(userSignUpRequest.getEncryptedPasswordUserSignUp(encryptor));
        }
    }

    public boolean isDuplicate(String id) {
        return userMapper.findById(id) != null;
    }

    public UserInfo getUserInfoById(String id) {
        return userMapper.findById(id);
    }

    public UserInfo getByIdAndPw(String id, String password) {
        UserInfo userInfo = userMapper.findByIdAndPassword(id, encryptor.encrypt(password));
        if (isValidUser(userInfo)) {
            return userInfo;
        } else {
            throw new NoSuchUserException("해당하는 사용자 정보가 없습니다.");
        }
    }

    private boolean isValidUser(UserInfo userInfo) {
        return userInfo != null;
    }

    public void updateById(String id, UserUpdateRequest userUpdateRequest) {
        //현재 비밀번호의 사용자가 존재하는지 확인
        getByIdAndPw(id, userUpdateRequest.getCurrentPassword());
        userMapper.updateById(id, encryptor.encrypt(userUpdateRequest.getNewPassword()),
                userUpdateRequest.getName(), userUpdateRequest.getEmail(),
                userUpdateRequest.getTelephone());
    }

    public void deleteById(String id) {
        if (isValidUser(userMapper.findById(id))) {
            userMapper.deleteById(id);
        } else {
            throw new NoSuchUserException("해당하는 사용자 정보가 없습니다.");
        }
    }

}

File: src/main/java/com/campool/service/UserService.kt
package com.campool.service

import com.campool.mapper.UserMapper
import com.campool.encrypt.Encryptor
import com.campool.model.UserSignUpRequest
import com.campool.model.UserInfo
import com.campool.exception.NoSuchUserException
import com.campool.model.UserUpdateRequest
import org.springframework.dao.DuplicateKeyException
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userMapper: UserMapper,
    private val encryptor: Encryptor
) {
    fun add(userSignUpRequest: UserSignUpRequest) {
        if (isDuplicate(userSignUpRequest.id)) {
            throw DuplicateKeyException("중복된 아이디가 존재합니다.")
        }

        userMapper.insertUser(userSignUpRequest.getEncryptedPasswordUserSignUp(encryptor))
    }

    fun isDuplicate(id: String): Boolean = userMapper.findById(id) != null

    fun getUserInfoById(id: String): UserInfo? = userMapper.findById(id)

    fun getByIdAndPw(id: String, password: String): UserInfo? {
        val userInfo = userMapper.findByIdAndPassword(id, encryptor.encrypt(password))
        return if (isValidUser(userInfo)) userInfo else throw NoSuchUserException("해당하는 사용자 정보가 없습니다.")
    }

    private fun isValidUser(userInfo: UserInfo?): Boolean = userInfo != null

    fun updateById(id: String, userUpdateRequest: UserUpdateRequest) {
        //현재 비밀번호의 사용자가 존재하는지 확인
        getByIdAndPw(id, userUpdateRequest.currentPassword)
        userMapper.updateById(
            id, encryptor.encrypt(userUpdateRequest.newPassword),
            userUpdateRequest.name, userUpdateRequest.email,
            userUpdateRequest.telephone
        )
    }

    fun deleteById(id: String) {
        if (!isValidUser(userMapper.findById(id))) {
            throw NoSuchUserException("해당하는 사용자 정보가 없습니다.")
        }

        userMapper.deleteById(id)
    }
}